[mono] Add `X86Base` support.

imhameed opened the pull request:
Implement `BitScanForward` and `BitScanReverse` as intrinsics.

Implement `__cpuidex` as a netcore-only icall. Rename `cpuid` (in
mono-hwcap-x86.c) to `mono_cpuidex`, give it extern linkage, and add
support for extended CPUID.

Related: https://github.com/dotnet/runtime/pull/40167

imhameed committed:
[mono] Add `X86Base` support.

Implement `BitScanForward` and `BitScanReverse` as intrinsics.

Implement `__cpuidex` as a netcore-only icall. Rename `cpuid` (in
mono-hwcap-x86.c) to `mono_cpuidex`, give it extern linkage, and add
support for extended CPUID.

Related: https://github.com/dotnet/runtime/pull/40167

imhameed committed:
extern private static -> private static extern

imhameed committed:
Define the __cpuidex icall even if DISABLE_SIMD is true.

jaykrell commented on src/mono/mono/mini/mini-llvm.c:
```diff
@@ -5712,6 +5712,28 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
 		}
+		case OP_X86_BSF32:
+		case OP_X86_BSF64: {
+			LLVMValueRef args [2] = {
```
No need for `2`.

CoffeeFlux commented on src/mono/mono/metadata/icall-decl.h:
```diff
@@ -304,4 +304,8 @@ ICALL_EXPORT gint32   ves_icall_System_Threading_LowLevelLifoSemaphore_TimedWait
 ICALL_EXPORT void     ves_icall_System_Threading_LowLevelLifoSemaphore_ReleaseInternal (gpointer sem_ptr, gint32 count);
 #endif
 
+#if defined(ENABLE_NETCORE)
```
Shouldn't this be behind `defined(TARGET_AMD64)` as well?

imhameed replied:
Yes.

imhameed committed:
Use 64-bit constants when xoring the result of a 64-bit ctlz.

Address feedback: remove the size from some local array definitions, and
hide ves_icall_System_Runtime_Intrinsics_X86_X86Base___cpuidex behind
TARGET_AMD64.

akoeplinger commented:
@imhameed this causes a new warning during the build:
```
  /Users/alexander/dev/runtime2/src/mono/mono/utils/mono-hwcap-x86.c:33:1: warning: no previous prototype for function 'mono_cpuidex' [-Wmissing-prototypes]
  mono_cpuidex (int id, int sub_id, int *p_eax, int *p_ebx, int *p_ecx, int *p_edx)
  ^
  /Users/alexander/dev/runtime2/src/mono/mono/utils/mono-hwcap-x86.c:32:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  gboolean
  ^
  static
  1 warning generated.
```
